contracts/wawaka/common/contract/attestation.cpp

```diff
@@ -0,0 +1,363 @@
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+bool ww::contract::attestation::get_ledger_key(
+    const Message& msg,
+    const Environment& env,
+    Response& rsp)
+{
+    ASSERT_SENDER_IS_CREATOR(env, rsp);
+
+    std::string ledger_verifying_key;
+    if (! get_ledger_key(ledger_verifying_key))
+        return rsp.error("failed to get the ledger verifying key");
+
+    ww::value::String result(ledger_verifying_key.c_str());
+    return rsp.value(result, false);
+}
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+// NAME: get_contract_metadata
+//
+// Contract method to get encryption and verifying key associated
+// with the contract.
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+bool ww::contract::attestation::get_contract_metadata(
+    const Message& msg,
+    const Environment& env,
+    Response& rsp)
+{
+    std::string verifying_key;
+    if (! KeyValueStore::privileged_get("ContractKeys.Verifying", verifying_key))
+        return rsp.error("failed to retreive privileged value for ContractKeys.Verifying");
+
+    std::string encryption_key;
+    if (! KeyValueStore::privileged_get("ContractKeys.Encryption", encryption_key))
+        return rsp.error("failed to retreive privileged value for ContractKeys.Encryption");
+
+    ww::value::Structure result(CONTRACT_METADATA_SCHEMA);
+    result.set_string("verifying_key", verifying_key.c_str());
+    result.set_string("encryption_key", encryption_key.c_str());
+
+    return rsp.value(result, false);
+}
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+// NAME: get_contract_code_metadata
+//
+// Contract method to get the code metadata associated with the
+// contract. Note that only the creator can get this information.
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+bool ww::contract::attestation::get_contract_code_metadata(
+    const Message& msg,
+    const Environment& env,
+    Response& rsp)
+{
+    ASSERT_SENDER_IS_CREATOR(env, rsp);
+
+    // returning all the information for now though only the nonce
+    // should be necessary for the attestation test
+    std::string code_nonce;
+    if (! KeyValueStore::privileged_get("ContractCode.Nonce", code_nonce))
+        return rsp.error("failed to retreive privileged value for ContractCode.Nonce");
+
+    ww::value::Structure result(CONTRACT_CODE_METADATA_SCHEMA);
+    result.set_string("code_nonce", code_nonce.c_str());
+
+    return rsp.value(result, false);
+}
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+// NAME: add_endpoint
+//
+// Contract method to verify the attestation of a contract object
+// and add it to the set of "trusted endpoints"
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+bool ww::contract::attestation::add_endpoint(
+    const Message& msg,
+    const Environment& env,
+    Response& rsp)
+{
+    // to add an endpoint we do need the ledger key
```

Comment:
Anybody can invoke this method. Is this intended?
It seems odd that anybody can create a contract and (being the creator) add it.

Reply:
More or less the same comment as above. It is fairly easy to wrap this with more "policy".